Core/src/org/sleuthkit/autopsy/centralrepository/eventlisteners/CaseEventListener.java

```diff
@@ -294,6 +298,97 @@ public void run() {
+            if (!EamDb.isEnabled()) {
+                return;
+            }
+            String modifiedTagName = (String) event.getNewValue();
+            List<String> notableTags = TagsManager.getNotableTagDisplayNames();
+            TskData.FileKnown status = notableTags.contains(modifiedTagName) ? TskData.FileKnown.BAD : TskData.FileKnown.UNKNOWN;
+            /**
+             * Set knownBad status for all files/artifacts in the given case
+             * that are tagged with the given tag name.
+             */
+            try {
+                TagName tagName = Case.getCurrentCase().getServices().getTagsManager().getDisplayNamesToTagNamesMap().get(modifiedTagName);
+                // First find any matching artifacts
+                List<BlackboardArtifactTag> artifactTags = Case.getCurrentCase().getSleuthkitCase().getBlackboardArtifactTagsByTagName(tagName);
+                for (BlackboardArtifactTag bbTag : artifactTags) {
+                    List<CorrelationAttribute> convertedArtifacts = EamArtifactUtil.getCorrelationAttributeFromBlackboardArtifact(bbTag.getArtifact(), true, true);
+                    for (CorrelationAttribute eamArtifact : convertedArtifacts) {
+                        boolean hasOtherBadTags = false;
+                        //if the new status of the tag is unknown UNKNOWN ensure we are not changing the status of BlackboardArtifact which still have other tags with a non-unknown status
+                        if (status == TskData.FileKnown.UNKNOWN) {
+                            Content content = bbTag.getContent();
+                            if ((content instanceof AbstractFile) && (((AbstractFile) content).getKnown() == TskData.FileKnown.KNOWN)) {
+                                break;
+                            }
+                            BlackboardArtifact bbArtifact = bbTag.getArtifact();
+                            TagsManager tagsManager = Case.getCurrentCase().getServices().getTagsManager();
+                            List<BlackboardArtifactTag> tags = tagsManager.getBlackboardArtifactTagsByArtifact(bbArtifact);
+                            for (BlackboardArtifactTag t : tags) {
+                                //avoid the possibility for threading issues if the tag whose status is currently changing is ever still in the tags manager with the old status
+                                if (t.getName().equals(tagName)) {
+                                    continue;
+                                }
+                                //if any other tags on this artifact are Notable in status then this artifact can not have its status changed 
+                                if (notableTags.contains(t.getName().getDisplayName())) {
+                                    hasOtherBadTags = true;
+                                    break;
+                                }
+                            }
+                        }
+                        if (!hasOtherBadTags) {
```

Comment:
I have not been able to figure out what this method is doing. It  seems possible that this code belongs outside of the loop, given that the flag is being set and used in the loop and the loop breaks if the flag is set. I'm just not sure though. Please add some comments to clarify the algorithm.

Reply:
This code is already outside of inner most for loop which the flag is set in and breaks out of upon being set.  More comments are incoming though